README.md
# tornado_server_status

CHANGELOG.md
# Changelog

## Development

## 0.1.0 (2022-03-19)

Initial release


tests/test_client_info.py
import pytest
import asyncio
import asyncssh

from tornado_server_status.client_info import run_cmdline_get_out

class TestClientInfo:
    hostname = '127.0.0.1'
    username = 'root'
    port = 22
    conn = None

    # @classmethod
    # @pytest.mark.asyncio
    # async def setup_class(cls):
    #     self.conn = await asyncssh.connect(self.hostname, username=self.username, \
    #                                 port=self.port, known_hosts=None)

    # @classmethod
    # @pytest.mark.asyncio
    # async def teardown_class(cls):
    #     await self.conn.close()

    # due to this issue, we cannt use cls.conn
    @pytest.mark.asyncio
    async def test_run_cmdline_get_out(self):
        # assert 1==0
        cmdline = 'echo 1'
        async with asyncssh.connect(self.hostname, username=self.username, \
                port=self.port, known_hosts=None) as conn:
            out = await run_cmdline_get_out(conn, cmdline)
            assert out.strip() == '1'


tornado_server_status/test.sh
_exists() {
    local cmd="$1"
    if eval type type > /dev/null 2>&1; then
        eval type "$cmd" > /dev/null 2>&1
    elif command > /dev/null 2>&1; then
        command -v "$cmd" > /dev/null 2>&1
    else
        which "$cmd" > /dev/null 2>&1
    fi
    local rt=$?
    return ${rt}
}
cname=$( awk -F: '/model name/ {name=$2} END {print name}' /proc/cpuinfo | sed 's/^[ \t]*//;s/[ \t]*$//' )
_exists "dmesg" && virtualx="$(dmesg 2>/dev/null)"
if _exists "dmidecode"; then
    sys_manu="$(dmidecode -s system-manufacturer 2>/dev/null)"
    sys_product="$(dmidecode -s system-product-name 2>/dev/null)"
    sys_ver="$(dmidecode -s system-version 2>/dev/null)"
else
    sys_manu=""
    sys_product=""
    sys_ver=""
fi
if   grep -qa docker /proc/1/cgroup; then
    virt="Docker"
elif grep -qa lxc /proc/1/cgroup; then
    virt="LXC"
elif grep -qa container=lxc /proc/1/environ; then
    virt="LXC"
elif [[ -f /proc/user_beancounters ]]; then
    virt="OpenVZ"
elif [[ "${virtualx}" == *kvm-clock* ]]; then
    virt="KVM"
elif [[ "${cname}" == *KVM* ]]; then
    virt="KVM"
elif [[ "${cname}" == *QEMU* ]]; then
    virt="KVM"
elif [[ "${virtualx}" == *"VMware Virtual Platform"* ]]; then
    virt="VMware"
elif [[ "${virtualx}" == *"Parallels Software International"* ]]; then
    virt="Parallels"
elif [[ "${virtualx}" == *VirtualBox* ]]; then
    virt="VirtualBox"
elif [[ -e /proc/xen ]]; then
    virt="Xen"
elif [[ "${sys_manu}" == *"Microsoft Corporation"* ]]; then
    if [[ "${sys_product}" == *"Virtual Machine"* ]]; then
        if [[ "${sys_ver}" == *"7.0"* || "${sys_ver}" == *"Hyper-V" ]]; then
            virt="Hyper-V"
        else
            virt="Microsoft Virtual Machine"
        fi
    fi
else
    virt="Dedicated"
fi
echo $virt


tornado_server_status/run_server_status.py
#!/usr/bin/env python
import os
import time
import logging
import pkgutil

import asyncio
from collections import defaultdict

import tornado.ioloop
import tornado.web
from tornado.options import define, options
from tornado.options import parse_command_line, parse_config_file

import asyncssh
import traceback

from .client_info import get_stats_data

import tornado.ioloop
import tornado.web

from pkg_resources import resource_string


BASE_DIR = os.path.dirname(os.path.abspath(__file__))


class APIhandler(tornado.web.RequestHandler):

    async def get_server_status(self, host, first_query=None):
        data = {}
        data['name'] = host
        data['type'] = 'Unkown'
        data['custom'] = ""
        data['location'] = 'Unkown'
        data['host'] = None
        data['online4'] = False
        data['online6'] = False
        data.update(self.application.first_query_results[host])
        try:
            conn = self.application.conns.get(host)
            if not conn is None:
                return_data = await asyncio.wait_for(
                    get_stats_data(conn, first_query=first_query),
                    timeout=options.ssh_query_timeout
                )
                data['online4'] = True
                data['online6'] = True
                data.update(return_data)
        except Exception as e:
            if options.debug:
                traceback.print_exc()
                logging.exception(e)
        return data

    async def get(self):
        # 尝试建立链接，如果有链接，则直接返回链接，否则
        # idea from https://github.com/ronf/asyncssh/issues/270
        servers = self.application.servers
        tasks = []
        for server in servers:
            host, username, port, password = '127.0.0.1', 'root', '22', None
            try:
                if len(server) == 1:
                    host = server
                elif len(server) == 2:
                    host, username = server
                elif len(server) == 3:
                    host, username, port = server
                elif len(server) == 4:
                    host, username, port, password = server
                conn = self.application.conns.get(host)
                if not conn:
                    last_failed_t = self.application.failed_hosts[host]
                    now = time.time()
                    interval = (now - last_failed_t)
                    if interval > options.ssh_connected_retry_interval:
                        conn = await asyncio.wait_for(
                            asyncssh.connect(
                                host=host, username=username, port=port,
                                password=password, known_hosts=None
                            ),
                            timeout=options.ssh_connected_timeout
                        )
                        self.application.conns[host] = conn
                        first_query_res = await self.get_server_status(
                                            host, first_query=True
                                        )
                        assert first_query_res['online4']
                        if first_query_res['online4']:
                            self.application.first_query_results[host] = first_query_res
                            print(host, 'connected!')
            except Exception as e:
                self.application.failed_hosts[host] = time.time()
                if options.debug:
                    logging.exception(e)

            tasks.append(
                self.get_server_status(host)
            )

        datas = await asyncio.gather(*tasks)
        res = {
            'servers': datas,
            'updated': f'{int(time.time())}'
        }
        self.write(res)


class BaseHandler(tornado.web.RequestHandler):
    def get_current_user(self):
        if options.password is None:
            return True
        return self.get_secure_cookie("user")


class MainHandler(BaseHandler):
    @tornado.web.authenticated
    def get(self):
        html_path = os.path.join(BASE_DIR, 'templates/web/index.html')
        self.render(html_path)


class LoginHandler(BaseHandler):

    def get(self):
        error_message = ""
        html_path = os.path.join(BASE_DIR, 'templates/web/login.html')
        self.render(html_path, error_message=error_message)

    def post(self):
        error_message = "Wrong name or password"
        username = self.get_argument("name", "")
        password = self.get_argument("password", "")
        if username == options.username and password == options.password:
            self.set_secure_cookie("user", username)
            self.redirect('/')
        else:
            html_path = os.path.join(BASE_DIR, 'templates/web/login.html')
            self.render(html_path, error_message=error_message)


class LogoutHandler(BaseHandler):
    def get(self):
        self.clear_all_cookies()
        self.redirect("/")


class ServerStatusApplication(tornado.web.Application):
    conns = defaultdict(lambda: None)
    failed_hosts = defaultdict(int)
    severs = []
    first_query_results = defaultdict(dict)

    def __init__(self, url_patterns,  **kwags):
        self.servers = options.servers
        tornado.web.Application.__init__(self, url_patterns, **kwags)


def main():
    # path to your settings module
    define("port", default=21388, help="run on the given port", type=int)
    define("config", default=None, help="tornado config file")
    define("debug", default=False, help="debug mode")
    define("servers", default=[('127.0.0.1', 'root', 22)], help="ssh connected timeout, default: 5.0")
    define("ssh_connected_timeout", default=5.0, help="ssh connect timeout, default: 5.0")
    define("ssh_connected_retry_interval", default=60.0, help="ssh connect retry interval, default: 60.0")
    define("ssh_query_timeout", default=5.0, help="ssh query timeout, default: 5.0")
    define("username", default=None, help="username")
    define("password", default=None, help="password")


    parse_command_line()
    if options.config:
        parse_config_file(options.config)
    if options.debug:
        tornado.log.enable_pretty_logging()

    static_dir = os.path.join(BASE_DIR, './templates/web')

    tornado_app = ServerStatusApplication(
        [
            (r'/ss_static/(.*)', tornado.web.StaticFileHandler, {"path": static_dir}),
            (r'/json/.*', APIhandler),
            (r"/login", LoginHandler),
            (r'/logout', LogoutHandler),
            (r'.*', MainHandler),
        ],
        debug=options.debug,
        autoreload=options.debug,
        login_url='/login',
        cookie_secret="Bj0aRCDg8fPyCNsA9Aub8i32U"
    )

    print(f'Runining on: http://localhost:{options.port}')
    tornado_app.listen(options.port, address='0')
    tornado.ioloop.IOLoop.current().start()

if __name__ == "__main__":
    main()


tornado_server_status/client_info.py
#!/usr/bin/env python3
#-*- coding: utf-8 -*-
"""
@author: h12345jack
@file: client_info.py
@time: 2022/03/21
"""

import os
import re
import time
import logging
from collections import defaultdict

import asyncio
import asyncssh

asyncssh.logging.set_log_level(logging.WARNING)

TIMEOUT = 100.0

async def run_cmdline_get_out(conn, cmdline, timeout=TIMEOUT):
    result = await asyncio.wait_for(conn.run(cmdline), timeout=None)
    return result.stdout


async def get_uptime(conn):
    cmdline = 'cat /proc/uptime'
    result = await run_cmdline_get_out(conn, cmdline)

    uptime = result.splitlines()[0].split('.', 2)
    uptime = int(uptime[0])
    days   = int(uptime/60.0/60.0/24.0)
    hours  = int(uptime/60.0/60.0)
    mins   = int(uptime/60.0 % 60)
    secs   = int(uptime % 60)

    if days > 0:
        return f'{days} 天'
    else:
        return f'{hours}:{mins}:{secs}'


async def get_time(conn):
    cmdline = 'cat /proc/stat'
    out = await run_cmdline_get_out(conn, cmdline)
    time_list = out.split(' ')[2:6]
    for i in range(len(time_list)):
        time_list[i] = int(time_list[i])
    return time_list


async def get_cpu(conn):
    # TODO change to multiple cpu 
    interval = 0.1
    x = await get_time(conn)
    await asyncio.sleep(interval)
    y = await get_time(conn)
    for i in range(len(x)):
        y[i] = (y[i] - x[i])
    t = y
    st = sum(t)
    if st == 0:
        st = 1
    result = 100-(t[-1]*100.00/st)
    return round(result, 1)


async def get_loadavg(conn):
    cmdline = 'python3 -c "import os;print(os.getloadavg())"'
    out = await run_cmdline_get_out(conn, cmdline)
    return eval(out)


async def get_liuliang(conn):
    net_in = 0
    net_out = 0
    cmdline = 'cat /proc/net/dev'
    out = await run_cmdline_get_out(conn, cmdline)

    for line in out.split('\n'):
        netinfo = re.findall('([^\s]+):[\s]{0,}(\d+)\s+(\d+)\s+(\d+)\s+(\d+)\s+(\d+)\s+(\d+)\s+(\d+)\s+(\d+)\s+(\d+)\s+(\d+)\s+(\d+)', line)
        if netinfo:
            if netinfo[0][0] == 'lo' or 'tun' in netinfo[0][0] \
                    or 'docker' in netinfo[0][0] or 'veth' in netinfo[0][0] \
                    or 'br-' in netinfo[0][0] or 'vmbr' in netinfo[0][0] \
                    or 'vnet' in netinfo[0][0] or 'kube' in netinfo[0][0] \
                    or netinfo[0][1] == '0' or netinfo[0][9] == '0':
                continue
            else:
                net_in += int(netinfo[0][1])
                net_out += int(netinfo[0][9])
    return net_in, net_out


async def get_memory(conn):
    cmdline = 'cat /proc/meminfo'
    out = await run_cmdline_get_out(conn, cmdline)

    re_parser = re.compile(r'^(?P<key>\S*):\s*(?P<value>\d*)\s*kB')
    memory_info = dict()
    for line in out.split('\n'):
        match = re_parser.match(line)
        if not match:
            continue
        key, value = match.groups(['key', 'value'])
        memory_info[key] = int(value)
    MemTotal = float(memory_info['MemTotal'])
    MemUsed = MemTotal-float(memory_info['MemFree'])-float(memory_info['Buffers'])-float(memory_info['Cached'])-float(memory_info['SReclaimable'])
    SwapTotal = float(memory_info['SwapTotal'])
    SwapFree = float(memory_info['SwapFree'])
    return int(MemTotal), int(MemUsed), int(SwapTotal), int(SwapFree)


async def get_hdd(conn):
    cmdline = 'df -Tlm --total -t ext4 -t ext3 -t ext2 -t reiserfs -t jfs -t ntfs -t fat32 -t btrfs -t fuseblk -t zfs -t simfs -t xfs'
    out = await run_cmdline_get_out(conn, cmdline)

    total = out.splitlines()[-1]
    used = total.split()[3]
    size = total.split()[2]
    return int(size), int(used)


async def get_net_speed(conn):
    netSpeed = {
        'netrx': [0.0],
        'nettx': [0.0],
        'clock': [0.0],
         'diff': [0.0],
        'avgrx': [0],
        'avgtx': [0]
    }

    cmdline = 'cat /proc/net/dev'
    for _ in range(3):
        out = await run_cmdline_get_out(conn, cmdline)
        net_dev = out.splitlines()
        avgrx = 0
        avgtx = 0
        for dev in net_dev[2:]:
            dev = dev.split(':')
            if "lo" in dev[0] or "tun" in dev[0] \
                    or "docker" in dev[0] or "veth" in dev[0] \
                    or "br-" in dev[0] or "vmbr" in dev[0] \
                    or "vnet" in dev[0] or "kube" in dev[0]:
                continue
            dev = dev[1].split()
            avgrx += int(dev[0])
            avgtx += int(dev[8])
        now_clock = time.time()
        netSpeed["diff"].append(now_clock - netSpeed["clock"][-1])
        netSpeed["clock"].append(now_clock)
        netSpeed["netrx"].append(int((avgrx - netSpeed["avgrx"][-1]) / netSpeed["diff"][-1]))
        netSpeed["nettx"].append(int((avgtx - netSpeed["avgtx"][-1]) / netSpeed["diff"][-1]))
        netSpeed["avgrx"].append(avgrx)
        netSpeed["avgtx"].append(avgtx)

    def return_average_value(v_l):
        # due to the first one is zero
        return sum(v_l) / (len(v_l) - 1)
    netSpeed["diff"]  = return_average_value(netSpeed['diff'])
    netSpeed["clock"] = return_average_value(netSpeed['clock'])
    netSpeed["netrx"] = return_average_value(netSpeed["netrx"])
    netSpeed["nettx"] = return_average_value(netSpeed["nettx"])
    netSpeed["avgrx"] = return_average_value(netSpeed["avgrx"])
    netSpeed["avgtx"] = return_average_value(netSpeed["avgtx"])

    return netSpeed


async def get_tupd(conn):
    '''
    tcp, udp, process, thread count: for view ddcc attack , then send warning
    :return:
    '''
    cmdline = "ss -t|wc -l"
    s = await run_cmdline_get_out(conn, cmdline)
    t = int(s[:-1])-1
    cmdline = "ss -u|wc -l"
    s = await run_cmdline_get_out(conn, cmdline)
    u = int(s[:-1])-1
    cmdline = "ps -ef|wc -l"
    s = await run_cmdline_get_out(conn, cmdline)
    p = int(s[:-1])-2
    cmdline = "ps -eLf|wc -l"
    s = await run_cmdline_get_out(conn, cmdline)
    d = int(s[:-1])-2
    return t, u, p, d



async def get_virt_type(conn):
    cmdline = """
_exists() {
    local cmd="$1"
    if eval type type > /dev/null 2>&1; then
        eval type "$cmd" > /dev/null 2>&1
    elif command > /dev/null 2>&1; then
        command -v "$cmd" > /dev/null 2>&1
    else
        which "$cmd" > /dev/null 2>&1
    fi
    local rt=$?
    return ${rt}
}
cname=$( awk -F: '/model name/ {name=$2} END {print name}' /proc/cpuinfo | sed 's/^[ \t]*//;s/[ \t]*$//' )
_exists "dmesg" && virtualx="$(dmesg 2>/dev/null)"
if _exists "dmidecode"; then
    sys_manu="$(dmidecode -s system-manufacturer 2>/dev/null)"
    sys_product="$(dmidecode -s system-product-name 2>/dev/null)"
    sys_ver="$(dmidecode -s system-version 2>/dev/null)"
else
    sys_manu=""
    sys_product=""
    sys_ver=""
fi
if   grep -qa docker /proc/1/cgroup; then
    virt="Docker"
elif grep -qa lxc /proc/1/cgroup; then
    virt="LXC"
elif grep -qa container=lxc /proc/1/environ; then
    virt="LXC"
elif [[ -f /proc/user_beancounters ]]; then
    virt="OpenVZ"
elif [[ "${virtualx}" == *kvm-clock* ]]; then
    virt="KVM"
elif [[ "${cname}" == *KVM* ]]; then
    virt="KVM"
elif [[ "${cname}" == *QEMU* ]]; then
    virt="KVM"
elif [[ "${virtualx}" == *"VMware Virtual Platform"* ]]; then
    virt="VMware"
elif [[ "${virtualx}" == *"Parallels Software International"* ]]; then
    virt="Parallels"
elif [[ "${virtualx}" == *VirtualBox* ]]; then
    virt="VirtualBox"
elif [[ -e /proc/xen ]]; then
    virt="Xen"
elif [[ "${sys_manu}" == *"Microsoft Corporation"* ]]; then
    if [[ "${sys_product}" == *"Virtual Machine"* ]]; then
        if [[ "${sys_ver}" == *"7.0"* || "${sys_ver}" == *"Hyper-V" ]]; then
            virt="Hyper-V"
        else
            virt="Microsoft Virtual Machine"
        fi
    fi
else
    virt="Dedicated"
fi
echo $virt
    """

    out = await run_cmdline_get_out(conn, cmdline)
    return out.strip()


async def get_ip_country(conn):
    py_code = "import json; import urllib.request; f=urllib.request.urlopen('http://ipinfo.io');jd=json.loads(f.read());f.close();print(jd['country'])"
    cmdline = f'python3 -c "{py_code}"'
    out = await run_cmdline_get_out(conn, cmdline)
    out = out.strip() if len(out.strip()) else 'Unknown'
    return out


async def get_stats_data(conn, first_query=True):
    tasks = []
    results = []

    data = defaultdict(int)

    tasks.append(get_cpu(conn))
    results.append('cpu')
    
    tasks.append(get_uptime(conn))
    results.append('uptime')

    tasks.append(get_loadavg(conn))
    results.append(('load_1', 'load_5', 'load_15'))

    tasks.append(get_memory(conn))
    results.append(('memory_total', 'memory_used', 'swap_total', 'swap_free'))

    tasks.append(get_hdd(conn))
    results.append(('hdd_total', 'hdd_used'))

    tasks.append(get_net_speed(conn))
    results.append('netSpeed')

    tasks.append(get_liuliang(conn))
    results.append(('network_in', 'network_out'))
    # # todo：兼容旧版本，下个版本删除ip_status
    data['ip_status'] = True
    # data['ping_10010'] = lostRate.get('10010') * 100
    # data['ping_189'] = lostRate.get('189') * 100
    # data['ping_10086'] = lostRate.get('10086') * 100

    # data['time_10010'] = pingTime.get('10010')
    # data['time_189'] = pingTime.get('189')
    # data['time_10086'] = pingTime.get('10086')
    tasks.append(get_tupd(conn))
    results.append(('tcp', 'udp', 'process', 'thread'))

    if first_query:
        tasks.append(get_virt_type(conn))
        results.append('type')

        tasks.append(get_ip_country(conn))
        results.append('location')


    R_s = await asyncio.gather(*tasks)
    for l, r in zip(results, R_s):
        if isinstance(l, tuple):
            for l_1, r_1 in zip(l, r):
                data[l_1] = r_1
        else:
            data[l] = r       
    data['load'] = f"{data['load_1'], data['load_5'], data['load_15']}"
    data['swap_used'] = data['swap_total'] - data['swap_free']
    data['network_rx'] = data['netSpeed'].get("netrx", 0.0)
    data['network_tx'] = data['netSpeed'].get("nettx", 0.0)

    return data


async def test():
    hostname = '127.0.0.1'
    username = 'root'
    port = 22
    async with asyncssh.connect(hostname, username=username, port=port, known_hosts=None) as conn:
        data = await get_stats_data(conn)
        print(data)


if __name__ == '__main__':
    asyncio.run(test())
